File: next-js-boilerplate/src/api/client/posts/actions.ts
import { useQueryClient } from "@tanstack/react-query";

export function usePostActions() {
  const queryClient = useQueryClient();

  const invalidate = () =>
    Promise.all([
      queryClient.invalidateQueries({ queryKey: ["feed"] }),
      queryClient.invalidateQueries({ queryKey: ["posts"] }),
    ]);

  const createPost = async (
    title: string,
    content: string,
    imageUrl?: string,
  ) => {
    const { createPostServer } = await import("@/api/server/posts/create");
    await createPostServer(title, content, imageUrl);
    await invalidate();
  };

  const updatePost = async (id: string, title: string, content: string) => {
    const { updatePostServer } = await import("@/api/server/posts/update");
    await updatePostServer(id, title, content);
    await invalidate();
  };

  const deletePost = async (id: string) => {
    const { deletePostServer } = await import("@/api/server/posts/delete");
    await deletePostServer(id);
    await invalidate();
  };

  const toggleReaction = async (params: {
    type: string;
    postId?: string;
    commentId?: string;
  }) => {
    const { toggleReactionServer } =
      await import("@/api/server/posts/reactions");
    await toggleReactionServer(params);
    await invalidate();
  };

  const createComment = async (
    postId: string,
    body: string,
    parentId?: string | null,
  ) => {
    const { createCommentServer } = await import("@/api/server/posts/comments");
    await createCommentServer(postId, body, parentId);
    await invalidate();
  };

  const updateComment = async (commentId: string, body: string) => {
    const { updateCommentServer } = await import("@/api/server/posts/comments");
    await updateCommentServer(commentId, body);
    await invalidate();
  };

  const deleteComment = async (commentId: string) => {
    const { deleteCommentServer } = await import("@/api/server/posts/comments");
    await deleteCommentServer(commentId);
    await invalidate();
  };

  const refreshPost = async (id: string) => {
    const { fetchSinglePostServer } = await import("@/api/server/posts/single");
    return fetchSinglePostServer(id);
  };

  return {
    createPost,
    updatePost,
    deletePost,
    toggleReaction,
    createComment,
    updateComment,
    deleteComment,
    refreshPost,
  };
}

File: flutter-boilerplate/lib/api/client/posts/actions.dart
import 'package:flutter_boilerplate/lib/riverpod_compat.dart';

import '../../../types/feed/comment.dart';
import '../../server/posts/comments.dart';
import '../../server/posts/create.dart';
import '../../server/posts/delete.dart';
import '../../server/posts/reactions.dart';
import '../../server/posts/update.dart';
import '../../server/posts/upload.dart';
import 'query.dart';

final postActionsProvider = Provider((ref) => PostActions(ref));

class PostActions {
  final Ref _ref;

  PostActions(this._ref);

  Future<void> create({
    required String title,
    required String content,
    String? imageUrl,
  }) async {
    final server = _ref.read(postCreateServerProvider);
    await server.call(title: title, content: content, imageUrl: imageUrl);
    _invalidateFeed();
  }

  Future<void> delete(String postId) async {
    final server = _ref.read(postDeleteServerProvider);
    await server.call(postId);
    _invalidateFeed();
    _ref.invalidate(postProvider(postId));
  }

  Future<void> update(
    String postId, {
    String? title,
    String? content,
    String? imageUrl,
  }) async {
    final server = _ref.read(postUpdateServerProvider);
    await server.call(
      postId,
      title: title,
      content: content,
      imageUrl: imageUrl,
    );
    _invalidateFeed();
    _ref.invalidate(postProvider(postId));
  }

  Future<void> toggleReaction(String postId, {String type = 'LIKE'}) async {
    final server = _ref.read(postReactionsServerProvider);
    await server.toggle(postId, type: type);
    _invalidateFeed();
    _ref.invalidate(postProvider(postId));
  }

  Future<Comment> addComment(
    String postId,
    String bodyText, {
    String? parentId,
  }) async {
    final server = _ref.read(postCommentsServerProvider);
    final comment = await server.create(postId, bodyText, parentId: parentId);
    _ref.invalidate(postCommentsProvider(comment.postId));
    _invalidateFeed();
    _ref.invalidate(postProvider(comment.postId));
    return comment;
  }

  Future<Comment> updateComment(
    String commentId, {
    required String bodyText,
  }) async {
    final server = _ref.read(postCommentsServerProvider);
    final comment = await server.update(commentId, bodyText: bodyText);
    _ref.invalidate(postCommentsProvider(comment.postId));
    _invalidateFeed();
    _ref.invalidate(postProvider(comment.postId));
    return comment;
  }

  Future<Comment> deleteComment(String commentId) async {
    final server = _ref.read(postCommentsServerProvider);
    final comment = await server.delete(commentId);
    _ref.invalidate(postCommentsProvider(comment.postId));
    _invalidateFeed();
    _ref.invalidate(postProvider(comment.postId));
    return comment;
  }

  Future<void> toggleCommentReaction(
    String commentId, {
    required String postId,
    String type = 'LIKE',
  }) async {
    final server = _ref.read(postReactionsServerProvider);
    await server.toggleForComment(commentId, type: type);
    _ref.invalidate(postCommentsProvider(postId));
    _invalidateFeed();
    _ref.invalidate(postProvider(postId));
  }

  Future<String> uploadImage(String filePath) async {
    final server = _ref.read(postUploadServerProvider);
    return server.call(filePath);
  }

  // Mirrors the web's usePostActions(), which invalidates every `["feed",
  // ...]`-keyed query after all 8 of these mutations, not just the comment
  // ones — every feed-list-shaped provider that could change as a result of
  // any of them, not just the one the calling screen happens to render.
  void _invalidateFeed() {
    _ref.invalidate(paginatedFeedProvider);
    _ref.invalidate(feedProvider);
  }
}
